icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1636,6 +1638,53 @@ private String replaceSubdivision(String subdivision) {
             return subdivisionAliasMap.get(subdivision);
         }
 
+        private String replaceTransformedExtensions(String extensions) {
+            StringBuilder builder = new StringBuilder();
+            List<String> subtags = new ArrayList<>(Arrays.asList(extensions.split(LanguageTag.SEP)));
+            List<String> tfields = new ArrayList<>();
+            int processedLength = 0;
+            int tlangLength = 0;
+            String tkey = "";
+            for (String subtag : subtags) {
+                if (LanguageTag.isTKey(subtag)) {
+                    if (tlangLength == 0) {
+                        tlangLength = processedLength-1;
```

Comment:
I find this Java code more tricky and obscure than the C++ version. I think it works, but please add comments. Here for example:
```
// Found the first tkey. Record the total length of the preceding tlang subtags.
// -1 if there is no tlang before the first tkey.
```